src/lib/index.ts

```diff
@@ -0,0 +1,272 @@
+import * as fs from 'fs';
+import * as mime from 'mime';
+import * as request from 'request';
+
+const gp12pem = require('google-p12-pem');
+const jws = require('jws');
+
+const GOOGLE_TOKEN_URL = 'https://accounts.google.com/o/oauth2/token';
+const GOOGLE_REVOKE_TOKEN_URL =
+    'https://accounts.google.com/o/oauth2/revoke?token=';
+
+export interface Payload {
+  iss: string;
+  scope: string|Array<string>;
+  aud: string;
+  exp: number;
+  iat: number;
+  sub: string;
+}
+
+export interface TokenOptions {
+  keyFile?: string;
+  key?: string|undefined;
+  email?: string|undefined;
+  iss?: string;
+  sub?: string;
+  scope?: string|Array<string>;
+}
+
+export class GoogleToken {
+  public token: string|null;
+  public expiresAt: number|null;
+  public key: string|undefined;
+  public keyFile: string|undefined;
+  public iss: string|undefined;
+  public sub: string;
+  public scope: string|undefined;
+  public rawToken: string|null;
+  public tokenExpires: number|null;
+  public email: string;
+
+  /**
+   * Create a GoogleToken.
+   *
+   * @param options  Configuration object.
+   */
+  constructor(options?: TokenOptions) {
+    this.configure(options);
+  }
+
+  /**
+   * Returns whether the token has expired.
+   *
+   * @return true if the token has expired, false otherwise.
+   */
+  public hasExpired() {
+    const now = (new Date()).getTime();
+    if (this.token && this.expiresAt) {
+      return now >= this.expiresAt;
+    } else {
+      return true;
+    }
+  }
+
+  /**
+   * Returns a cached token or retrieves a new one from Google.
+   *
+   * @param callback The callback function.
+   */
+  public getToken(callback: (err: Error|null, token?: string|null) => void) {
+    const handleJSONKey = (err: Error, key: string) => {
+      if (err) {
+        callback(err);
+        return;
+      }
+      try {
+        const body = JSON.parse(key);
+        this.key = body.private_key;
+        this.iss = body.client_email;
+      } catch (e) {
+        callback(e);
+        return;
+      }
+
+      if (!this.key || !this.iss) {
+        const error = new Error('private_key and client_email are required.');
+        (<NodeJS.ErrnoException>error).code = 'MISSING_CREDENTIALS';
```

Comment:
nit (but this can be addressed in a follow on once we pick up https://github.com/google/ts-style/pull/75). Prefer `(.. as Type)` casts over angle bracket versions.

Reply:
fixed